readthedocs/projects/models.py

```diff
@@ -1465,14 +1469,105 @@ def __str__(self):
         return self.email
 
 
-class WebHook(Notification):
+class WebHookEvent(models.Model):
+
+    BUILD_TRIGGERED = 'build:triggered'
+    BUILD_PASSED = 'build:passed'
+    BUILD_FAILED = 'build:failed'
+
+    EVENTS = (
+        (BUILD_TRIGGERED, _('Build triggered')),
+        (BUILD_PASSED, _('Build passed')),
+        (BUILD_FAILED, _('Build failed')),
+    )
+
+    name = models.CharField(
+        max_length=256,
+        unique=True,
+        choices=EVENTS,
+    )
+
+    def __str__(self):
+        return self.name
+
+
+class WebHook(Notification, TimeStampedModel):
+
     url = models.URLField(
+        _('URL'),
         max_length=600,
         help_text=_('URL to send the webhook to'),
     )
+    secret = models.CharField(
+        help_text=_('Secret used to sign the payload of the webhook'),
+        max_length=255,
+        blank=True,
+        null=True,
+    )
+    events = models.ManyToManyField(
+        WebHookEvent,
+        related_name='webhooks',
+        help_text=_('Events to subscribe'),
+    )
+    payload = models.TextField(
+        _('JSON payload'),
+        help_text=_(
+            'JSON payload to send to the webhook. '
+            'Check the docs for available substitutions.',
+        ),
+        blank=True,
+        null=True,
+    )
+    exchanges = GenericRelation(
+        'integrations.HttpExchange',
+        related_query_name='webhook',
+    )
+
+    SUBSTITUTIONS = (
+        ('${event}', _('Event that triggered the request')),
+        ('${build.id}', _('Build ID')),
+        ('${build.commit}', _('Commit being built')),
+        ('${project.slug}', _('Project slug')),
+        ('${project.name}', _('Project name')),
+        ('${version.id}', _('Version ID')),
+        ('${version.slug}', _('Version slug')),
```

Comment:
Isn't it better to use Django template substitutions here and render the template instead of using our custom method?

Reply:
Two things:

- security
- feature bloat

If users can write their own template, they could access to things we don't want to. Even if we pass plain values to the context, I think we still pass a global context to all templates.
And we will be giving access to filters, blocks, directives, etc.

Even if we use python's `.format`, that still gives users access to additional format options like padding and what not. We don't want to support that, a simple search and replace should be enough.

Reply:
We can full control over what we pass to `Template.render()`, so we can avoid passing global context. However, I understand that filter and blocks could be used by the user and we may not want that.

That said, I still think that it would be good to use a standardized way for substitution (`{{ variable }}`)  instead of our custom one (`${variable}`). Many people are already familiar with the Django Template syntax and also it will keep the door open for the future to use `Template.render()` with a simplified `Engine` (https://docs.djangoproject.com/en/3.2/ref/templates/api/#django.template.Engine) that only supports variable substitution if we decide to go that path.

Reply:
I'm not sure if we want to expand this in the future TBH, if users want to have more granular control over the payload, they should use a proxy instead of solving that with a template solution.

I don't really like the django's template format, I'd prefer the python's format if we were to change this (and python's format is even more known). Mainly, the spaces around the brackets can be a source of confusion (like, are they always required? is it just one space needed? are tabs/nbsp/vertical tabs or other "space" chars allowed?)

Reply:
> I don't really like the django's template format, I'd prefer the python's format if we were to change this

Python format works better than a Read the Docs' custom one. However, I still prefer using a syntax that can be compatible with a known template engine (`{{ variable }}` works for Django Template and Jinja) because it gives us a lot of flexibility and security for the future.

> Mainly, the spaces around the brackets can be a source of confusion (like, are they always required? is it just one space needed? are tabs/nbsp/vertical tabs or other "space" chars allowed?)

The same questions and confusions apply to the format you are proposing. The difference is that using the Django engines solves this for you already 😄: space around the variable name is not required; both work.

Reply:
but we don't need to use a template engine, this is merely a search/replace operation.

>  because it gives us a lot of flexibility and security for the future.

I think that introduces a vector attack actually, a bug or misconfiguration on the template engine could lead to global access of our app.